Replace EnabledIf with RELEASE_ASSERT for ModelProcessModelPlayer and PlatformXRSystemProxy

szewai opened the pull request:
#### 306ff2dc43802343bd0b536d19c45b823807694e
<pre>
Replace EnabledIf with RELEASE_ASSERT for ModelProcessModelPlayer and PlatformXRSystemProxy
<a href="https://bugs.webkit.org/show_bug.cgi?id=283438">https://bugs.webkit.org/show_bug.cgi?id=283438</a>
<a href="https://rdar.apple.com/140295541">rdar://140295541</a>

Reviewed by Ryosuke Niwa.

It does not make much sense to add message check in web process since its main goal is to kill sender web process. In
this case, the message is sent by a trusted process and it is more likely something is wrong with the receiver web
process. This patch repalces them with RELEASE_ASSERT.

* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
(WebKit::ModelProcessModelPlayer::didCreateLayer):
(WebKit::ModelProcessModelPlayer::didFinishLoading):
(WebKit::ModelProcessModelPlayer::didUpdateEntityTransform):
(WebKit::ModelProcessModelPlayer::didUpdateAnimationPlaybackState):
(WebKit::ModelProcessModelPlayer::didFinishEnvironmentMapLoading):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.messages.in:
* Source/WebKit/WebProcess/XR/PlatformXRSystemProxy.cpp:
(WebKit::PlatformXRSystemProxy::sessionDidEnd):
(WebKit::PlatformXRSystemProxy::sessionDidUpdateVisibilityState):
* Source/WebKit/WebProcess/XR/PlatformXRSystemProxy.messages.in:

Canonical link: <a href="https://commits.webkit.org/286990@main">https://commits.webkit.org/286990@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5545e65445abf666a6a8a76cc638e2ba5c63f029

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77838 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56871 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/31223 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/82486 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29095 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66029 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5165 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60956 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18898 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80906 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50935 "Found 2 new test failures: fast/forms/date/date-input-rendering-basic.html fast/forms/date/date-pseudo-elements.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66829 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41259 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48290 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/27438 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69405 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24661 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83848 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5205 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/3504 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69178 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5361 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66757 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68427 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17098 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12436 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/10536 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/99386 "Build is in progress. Recent messages:") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5153 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/7906 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/35/builds/99386 "Build is in progress. Recent messages:") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5145 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8577 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6930 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
```diff
@@ -77,38 +79,48 @@ bool ModelProcessModelPlayer::modelProcessEnabled() const
+void ModelProcessModelPlayer::didCreateLayer(IPC::Connection& connection, WebCore::LayerHostingContextIdentifier identifier)
 {
     RELEASE_LOG(ModelElement, "%p - ModelProcessModelPlayer obtained new layerHostingContextIdentifier id=%" PRIu64, this, m_id.toUInt64());
+    MESSAGE_CHECK(modelProcessEnabled(), connection);
```
What's the point of a MESSAGE_CHECK in the WebProcess? We usually use those in trusted processes, when validating IPC from untrusted processes.

szewai replied:
Technically this is not as useful as in privileged processes. We do print log if a message is received unexpectedly though, so this might be used to find bugs in trusted process (e.g. feature-flag-syncing issue), and I just decided to not change the behavior in this patch.

darryl-apple replied:
Is there any benefit of removing EnabledIf's for cases like this? It's easier to see how message endpoints are guarded if all the logic is contained in the `.messages.in` files

cdumez replied:
@szewai But MESSAGE_CHECK doesn't seem to make sense in the WebProcess at all. MESSAGE_CHECK terminates the IPC connection normally. Not sure what MESSAGE_CHECK does in the WebProcess. Seems we'd get more actionable crashes with a RELEASE_ASSERT. MESSAGE_CHECK is meant to kill the compromised remote process which doesn't really apply here.

szewai replied:
I think you are making a good point about WCP could be killed for bug; I will just change it to RELEASE_ASSERT

eddydas replied:
Do we have an answer for Darryl's question? Like do we NEED to remove EnabledIf in .messages.in files?

szewai replied:
Yes, we are going to drop `EnabledIf`.
`EnabledIf` currently overlaps with `EnabledBy` and `MESSAGE_CHECK`, and we are going to drop `EnabledIf` in favor of `EnabledBy` and `MESSAGE_CHECK`.
For feature flag protection (i.e. message endpoint enablement based on runtime feature flag), we should use `EnabledBy`; for others (validating parameters or status of message receiver object),  we should use `MESSAGE_CHECK` in message handler functions. (Web process is an exception since it cannot kill other privileged process for invalid message, so we just use `RELEASE_ASSERT` instead of `MESSAGE_CHECK`.)
We are going to enforce `EnabledBy` annotation on new message endpoints soon. (We drop `EnabledIf` to avoid confusion on the requirement of feature flag protection).

szewai committed:
Replace EnabledIf with RELEASE_ASSERT for ModelProcessModelPlayer and PlatformXRSystemProxy
https://bugs.webkit.org/show_bug.cgi?id=283438
rdar://140295541

Reviewed by Ryosuke Niwa.

It does not make much sense to add message check in web process since its main goal is to kill sender web process. In
this case, the message is sent by a trusted process and it is more likely something is wrong with the receiver web
process. This patch repalces them with RELEASE_ASSERT.

* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
(WebKit::ModelProcessModelPlayer::didCreateLayer):
(WebKit::ModelProcessModelPlayer::didFinishLoading):
(WebKit::ModelProcessModelPlayer::didUpdateEntityTransform):
(WebKit::ModelProcessModelPlayer::didUpdateAnimationPlaybackState):
(WebKit::ModelProcessModelPlayer::didFinishEnvironmentMapLoading):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.messages.in:
* Source/WebKit/WebProcess/XR/PlatformXRSystemProxy.cpp:
(WebKit::PlatformXRSystemProxy::sessionDidEnd):
(WebKit::PlatformXRSystemProxy::sessionDidUpdateVisibilityState):
* Source/WebKit/WebProcess/XR/PlatformXRSystemProxy.messages.in:

Canonical link: https://commits.webkit.org/286990@main
